Fix rustdocs in `units` crate

jamillambert opened the pull request:
The rustdocs in the `units` crate has been reviewed and improved.

Some of the links were missing backticks, these have been added.

Some grammatical changes have been made to improve the documentation.

The use of links was inconsistent and has been changed to have links everywhere that seemed appropriate.

A couple of error headings were added and a description as to why a capital M is not accepted in the denomination for MegaSatoshi or MegaBitcoin.

Kixunil commented on units/src/amount.rs:
```diff
@@ -69,7 +69,7 @@ impl Denomination {
     }
 
-    /// Returns stringly representation of this
+    /// Returns a string representation of this amount.
```
denomination, not amount.

jamillambert replied:
Fixed

storopoli commented on units/src/amount.rs:
```diff
@@ -107,15 +107,25 @@ impl fmt::Display for Denomination {
+    ///
+    /// # Errors
+    ///
+    /// - If the denomination begins with a capital `M` a `PossiblyConfusingDenominationError` is
```
Is it possible to doclink?

```suggestion
    /// - If the denomination begins with a capital `M` a [`PossiblyConfusingDenominationError`] is
```

jamillambert replied:
Done

Kixunil commented on units/src/amount.rs:
```diff
@@ -107,7 +107,7 @@ impl fmt::Display for Denomination {
     type Err = ParseDenominationError;
 
-    /// Convert from a str to Denomination.
+    /// Converts from a str to a Denomination.
```
It'd be more appropriate to put the whole doc do on top of the trait impl, not method.

jamillambert replied:
I moved the section about Mega and accepted denominations to the top

tcharding replied:
We should backtick `str` and `Denomination` I believe (or use "denomination").

jamillambert replied:
backticks added

storopoli commented on units/src/amount.rs:
```diff
@@ -107,15 +107,25 @@ impl fmt::Display for Denomination {
+    /// - If the denomination begins with a capital `M` a `PossiblyConfusingDenominationError` is
+    /// returned.
+    ///
+    /// - If an unknown denomination is used, an `UnknownDenominationError` is returned.
```
Same here?

```suggestion
    /// - If an unknown denomination is used, an [`UnknownDenominationError`] is returned.
```

storopoli commented on units/src/amount.rs:
```diff
@@ -514,7 +524,7 @@ fn is_too_precise(s: &str, precision: usize) -> Option<usize> {
 
-/// Parse decimal string in the given denomination into a satoshi value and a
+/// Parses a decimal string in the given denomination into a satoshi value and a
 /// bool indicator for a negative amount.
```
I don't know if this is what we are doing now, but since `bool` is a Rust type we can backtick it right?

```suggestion
/// `bool` indicator for a negative amount.
```

storopoli replied:
maybe doclink as well?

jamillambert replied:
Done

Kixunil commented on units/src/amount.rs:
```diff
@@ -1381,45 +1381,56 @@ impl SignedAmount {
-    /// Get the absolute value of this [`SignedAmount`].
+    /// Returns the absolute value of this [`SignedAmount`].
+    ///
     /// Returns [`None`] if overflow occurred. (`self == MIN`)
```
You could add: "Consider using `unsigned_abs` which is often more practical."

jamillambert replied:
Added

Kixunil commented on units/src/parse.rs:
```diff
@@ -225,8 +225,8 @@ pub fn hex_u32_unprefixed(s: &str) -> Result<u32, UnprefixedHexError> {
-/// - With `InvalidDigit` due to the `x` if there is a prefix.
+/// - If the input string contains a `0x` (or `0X`) prefix, with `InvalidDigit` due to the `x`.
+/// - If the input string is not a valid hex encoding of a `u32`.
 pub fn hex_u32_unchecked(s: &str) -> Result<u32, ParseIntError> {
```
Hmm, why is it call unchecked? I don't see anything that needs checking and the doc doesn't explain it.

jamillambert replied:
It is used by the `hex_u32_unprefixed()` and `hex_u32_prefixed()` after first checking for a prefix.  i.e. compared to these two functions it is "unchecked".  I added to the doc that it doesn't first check for a prefix to try to clarify this.

Kixunil replied:
Looks like they should've been private...

jamillambert replied:
That's what I thought when I looked at  it just then.  I also just changed it for u128 too.

tcharding replied:
We use `hex_u128_unchecked` in `bitcoin::pow` so for API uniformity it seems correct to have this function public.

Kixunil replied:
@tcharding that sounds really weird, I'll need to investigate.

Kixunil commented on units/src/weight.rs:
```diff
@@ -123,7 +123,7 @@ impl Weight {
+    /// Scales by the witness factor.
     ///
     /// Computes `self * WITNESS_SCALE_FACTOR` returning `None` if an overflow occurred.
     pub fn scale_by_witness_factor(self) -> Option<Self> {
```
Why the hell do we have this function here? Weight already accounts for witness, so this is wrong. It should've taken `usize` and be called `from_non_discounted_size` (and also have `from_discounted_size`).

tcharding replied:
@Kixunil its not clear to me from your comment what the priority is with this problem. If you really care about something but don't have the time to go further can you mention me please and ask me to investigate it? (I think its fine to make requests of my time because I'm paid to be here, for others I don't think we should do it too much). Cheers man. (And well handled Jamil, we are a prickly bunch round here, we have a tendency to get a bit emotionally involved in our work, don't take it personally.)

Kixunil replied:
@tcharding thanks! Yes, that'd be appreciated. The priority is: I don't want to stabilize `units` with this function.

tcharding replied:
No worries, added #2966

storopoli commented on units/src/parse.rs:
```diff
@@ -377,7 +376,7 @@ impl From<ParseIntError> for UnprefixedHexError {
 }
 
-/// Error when hex string is missing a prefix (e.g. 0x).
+/// Error returned when a hex string is missing a prefix (e.g. 0x).
```
```suggestion
/// Error returned when a hex string is missing a prefix (e.g. `0x`).
```

jamillambert replied:
Done, and thanks for the review

storopoli reviewed (commented):
Some suggestions.
Nice work!

jamillambert commented:
Addressed review comments, split out "Add missing links" into its own commit which was accidentally combined with another commit previously.

tcharding commented on units/src/locktime/absolute.rs:
```diff
@@ -12,7 +12,8 @@ use crate::parse::ParseIntError;
 
-/// The Threshold for deciding whether a lock time value is a height or a time (see [Bitcoin Core]).
+/// The Threshold for deciding whether a lock time value is a height or a time
+/// (see [`Bitcoin Core`]).
```
I don't think this one is correct. From discussion just recently with @storopoli we are backticking anything that is typically typed as code but "Bitcoin Core" is the name of the project not a code term.

jamillambert replied:
Backticks removed

tcharding commented on units/src/amount.rs:
```diff
@@ -1366,61 +1384,74 @@ impl SignedAmount {
 
-    /// Returns `true` if this [SignedAmount] is positive and `false` if
-    /// this [SignedAmount] is zero or negative.
+    /// Checks if this [`SignedAmount`] is positive
```
```suggestion
    /// Checks if this [`SignedAmount`] is positive.
```

jamillambert replied:
Done

tcharding commented on units/src/amount.rs:
```diff
@@ -1366,61 +1384,74 @@ impl SignedAmount {
 
-    /// Returns `true` if this [SignedAmount] is negative and `false` if
-    /// this [SignedAmount] is zero or positive.
+    /// Checks if this [`SignedAmount`] is negative
```
```suggestion
    /// Checks if this [`SignedAmount`] is negative.
```

jamillambert replied:
Done

tcharding commented on units/src/parse.rs:
```diff
@@ -214,19 +214,20 @@ pub fn hex_u32_prefixed(s: &str) -> Result<u32, PrefixedHexError> {
 ///
-/// - If input string is not a valid hex encoding of a `u32`.
-/// - With `InvalidDigit` due to the `x` if there is a prefix.
+/// - If the input string contains a `0x` (or `0X`) prefix an `InvalidDigit` error is returned due
+///   to the `x`.
+/// - If the input string is not a valid hex encoding of a `u32`.
```
The improved clarity is good but the introduced line wrapping is a bit ugly, can we make it more terse? Perhaps

```rust
/// - If the input string contains a `0x` (or `0X`) prefix, returns `InvalidDigit` due to the `x`.
```

jamillambert replied:
Thanks, I have used your suggestion

tcharding commented on units/src/parse.rs:
```diff
@@ -252,8 +253,8 @@ pub fn hex_u128(s: &str) -> Result<u128, ParseIntError> {
 /// # Errors
 ///
-/// - If input string does not contain a `0x` (or `0X`) prefix.
-/// - If input string is not a valid hex encoding of a `u128`.
+/// - If the input string does not contain a `0x` (or `0X`) prefix.
+/// - If the input string is not a valid hex encoding of a `u128`.
```
I'm fine with this change but I think the original was correct too. Mentioning it because we discussed this before because it is hard  for English-second-language speakers to know which is correct.

tcharding commented on units/src/parse.rs:
```diff
@@ -263,19 +264,20 @@ pub fn hex_u128_prefixed(s: &str) -> Result<u128, PrefixedHexError> {
 }
 
-/// Parses a `u128` from a hex string.
+/// Parses a `u128` an unprefixed hex string without first checking for a prefix.
```
```suggestion
/// Parses a `u128` from an unprefixed hex string without first checking for a prefix.
```

jamillambert replied:
Done

tcharding commented on units/src/parse.rs:
```diff
@@ -263,19 +264,20 @@ pub fn hex_u128_prefixed(s: &str) -> Result<u128, PrefixedHexError> {
 ///
-/// - If input string is not a valid hex encoding of a `u128`.
-/// - With `InvalidDigit` due to the `x` if there is a prefix.
+/// - If the input string contains a `0x` (or `0X`) prefix an `InvalidDigit` error is returned due
```
Note, will need changing if you choose to change the docs on `hex_u32_unchecked`.

tcharding commented on units/src/weight.rs:
```diff
@@ -105,27 +105,27 @@ impl Weight {
     pub fn checked_div(self, rhs: u64) -> Option<Self> { self.0.checked_div(rhs).map(Self) }
 
-    /// Scale by witness factor.
+    /// Scales by the witness factor.
```
I prefer the original because the witness factor is not really a well known thing. I could be wrong, grammar is definitely not my strong suit.

jamillambert replied:
Removed "the"

tcharding commented on units/src/block.rs:
```diff
@@ -6,7 +6,7 @@
 //! These are general types for abstracting over block heights, they are not designed to use with
 //! lock times. If you are creating lock times you should be using the
-//! [`locktime::absolute::Height`] and [`locktime::relative::Height`] types.
+//! [`locktime::absolute::Height`](crate::locktime) and [`locktime::relative::Height`](crate::locktime) types.
```
This `(crate::locktime)` shouldn't be needed, we bring `locktime` into scope already using `#[cfg(doc)]`.

jamillambert replied:
Thanks, I see this now and have removed it.  I was running `cargo doc` locally to check the document pages and the link didn't work, I assume it is something obvious I am not doing like adding a needed option.

Also `pub struct BlockHeight` and `BlockInterval` don't appear on the `docs.rs` page so I couldn't check there.  The reason for this may also be obvious but why is that?

storopoli replied:
You might be missing the `RUSTDOCFLAGS` here https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/blob/3494ceec521cbc2cbe7e4b0d3fb9219a2164f57d/ci/run_task.sh#L277

@tcharding do you think it useful if we add a `just` command to build the docs?

```bash
just docs
```

jamillambert replied:
Thanks, that fixed it

tcharding replied:
I thought we added that `just` command but it seems not. I think we should, I personally use a shell alias `RUSTDOCFLAGS="--cfg docsrs" cargo +nightly rustdoc --all-features -- -D rustdoc::broken-intra-doc-links
`

tcharding replied:
I  created https://github.com/rust-bitcoin/rust-bitcoin/issues/2965

tcharding reviewed (commented):
Thanks for sticking with this @jamillambert, like Andrew said once before PRs like this really make the code look more professional and help with the imminent stable release.

tcharding commented:
Its not super important but we tend to not treat crate names as propper nouns, so we tend to write `units` instead of Units - at least I believe that is not just me.

jamillambert commented:
> Its not super important but we tend to not treat crate names as propper nouns, so we tend to write `units` instead of Units - at least I believe that is not just me.

Thanks, I changed the PR heading.  And I think I have addressed all of your comments.

tcharding commented on units/src/locktime/absolute.rs:
```diff
@@ -12,7 +12,8 @@ use crate::parse::ParseIntError;
 use crate::prelude::{Box, String};
 
-/// The Threshold for deciding whether a lock time value is a height or a time (see [Bitcoin Core]).
+/// The Threshold for deciding whether a lock time value is a height or a time
+/// (see [Bitcoin Core]).
```
This change shouldn't be in here now :)

tcharding reviewed (commented):
Almost there man, thanks for sticking with this, I know how it feels to continually rebase and fix large docs PRs , its hard yakka.

(Just a heads up, you are going to get merge conflicts with #2964.)

Kixunil commented on units/src/amount.rs:
```diff
@@ -109,13 +122,12 @@ impl FromStr for Denomination {
+    ///   returned.
     ///
-    /// Due to ambiguity between mega and milli we prohibit usage of leading capital 'M'.
+    /// - If an unknown denomination is used, an `UnknownDenominationError` is returned.
```
This whole doc should be moved on top of `impl` block to be more visible. Nobody will see it on a trait method.

But I don't really want to review this again, so I'd prefer it done in a followup PR.

jamillambert replied:
Sorry I had to remove the line wrap change and lost your approval.  I have not made any other changes and will look into the suggested doc change seperately.

Kixunil reviewed (approved):
ACK 2e5f58c2e2955abed1c62bd69ba84f35cc53875e

jamillambert commented:
> This change shouldn't be in here now :)

Thanks I missed this when removing the backticks.  I have removed the changed.

jamillambert commented:
Removed merge conflict with #2964

apoelstra reviewed (dismissed):
ACK 78486a67a1228bea3194b9a6aec8716fe4fbd91a

Kixunil reviewed (dismissed):
ACK 78486a67a1228bea3194b9a6aec8716fe4fbd91a

Kixunil commented on units/src/locktime/relative.rs:
```diff
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: CC0-1.0
 
-//! Provides type `Height` and `Time` types used by the `rust-bitcoin` `relative::LockTime` type.
+//! Provides [`Height`] and [`Time`] types used by the `rust-bitcoin` [`relative::LockTime`](crate::locktime) type.
```
Why is `relative::LockTime` not in this crate?

tcharding replied:
Because I argued that its not a unit but a primitive, IIRC @apoelstra was not too fussed either way. Less philosophically see the list of imports in `bitcoin::locktime::absolute` and `bitcoin::ocktime::relative`, they cannot be trivially moved to `units`.

Kixunil replied:
IMO it's more like `Sequence` is the primitive. But Let's keep it as-is, we can move it any time we want.

jamillambert replied:
Removed `relative::LockTime` links from relative.rs and absolute.rs.

apoelstra replied:
Agreed philosophically that `Sequence` is a primitive, `relative::LockTime` is a unit ... and as for `absolute::LockTime` who can say?

But as I recall the decision was actually made pragmatically because of difficulties moving the types.

Anyway this is not the place to litigate this :).

tcharding replied:
I'll comment for Jamil "STOP DERAILING MY THREAD" :)

apoelstra reviewed (approved):
ACK 6dd5af9678c40cac9e360dfcbd9623cdb774c755

tcharding reviewed (approved):
ACK 6dd5af9678c40cac9e360dfcbd9623cdb774c755

tcharding commented:
BOOM! Thanks for sticking with this one @jamillambert
